stan/math/rev/fun/quad_form.hpp

```diff
@@ -93,6 +93,65 @@ class quad_form_vari : public vari {
 
   quad_form_vari_alloc<Ta, Ra, Ca, Tb, Rb, Cb>* impl_;
 };
+
+/**
+ * Return the quadratic form \f$ B^T A B \f$.
+ *
+ * Symmetry of the resulting matrix is not guaranteed due to numerical
+ * precision.
+ *
+ * @tparam T1 type of the first (square) matrix
+ * @tparam T2 type of the second matrix
+ *
+ * @param A square matrix
+ * @param B second matrix
+ * @return The quadratic form, which is a symmetric matrix.
+ * @throws std::invalid_argument if A is not square, or if A cannot be
+ * multiplied by B
+ */
+template <typename Mat1, typename Mat2,
+          require_all_matrix_t<Mat1, Mat2>* = nullptr,
+          require_any_var_matrix_t<Mat1, Mat2>* = nullptr>
+inline auto quad_form_impl(const Mat1& A, const Mat2& B, bool symmetric) {
+  check_square("quad_form", "A", A);
+  check_multiplicable("quad_form", "A", A, "B", B);
+
+  using return_t
+    = promote_var_matrix_t<decltype(value_of(B).transpose().eval() *
+				    value_of(A) *
+				    value_of(B).eval()), Mat1, Mat2>;
+
+  auto arena_A = to_arena(A);
+  auto arena_B = to_arena(B);
```

Comment:
If these matrices are either `var<mat>` s or just regular `mat` s then this code is fine. If one of these is a `mat<var>` though then it's not as fine because I access values and adjoints repeatedly. If the vals/adjoints are accessed more than once, it would be more economical to just copy to a `var<mat>`. 

So my question is we should just by default be copying to `var<mat>` in all these functions?

Reply:
How about you do something like:
```
const auto& A_adj = to_ref(arena_A)
```
Also it would be nice if we only copied these matrices to arena if we actually needed them in the reverse pass.

Reply:
Well I'm not worried about evaluatiang expressions as much as converting things to `var<mat>` land so I can be lazy.

I checked and all the copying here is fine if inputs are a mix of `mat<double>` and `var<mat>`. The `A` inputs and the `B` inputs are required for both reverse passes. It's not so good for `mat<var>`.

Reply:
I see. `to_arena` does not convert to `var<mat>`, so there is steel a need to handle `mat<var>` efficiently.

Reply:
I guess the solution is to just add a version of `to_var_value` (https://github.com/stan-dev/math/blob/develop/stan/math/rev/fun/to_var_value.hpp) that just forwards things through if it is given a `var_value`.

Reply:
Oh it would also have to just pass through prims, and that's probably not what that does. `to_var_value(mat_double)` seems like it should create a `var<mat>` not just pass through a double matrix.

Reply:
Could do `to_var_value_if`, that's a pretty common sorta pattern around these parts.

Reply:
The conversion between `var<mat>` and `mat<var>` is quite expensive. So this would be quite lazy and inefficient option.

Reply:
Well it's only as expensive as a reading a `mat<var>` once, and so if the function uses the `mat<var>` more than once or twice we're probably good right?

Reply:
No. It is quite a bit more expensive, as it needs to also propagate adjoints back. Even we used a matrix many times manually copying values out and adjoints back in `mat<var>` would be cheaper.

Reply:
Copying values out and putting adjoints back in all `to_var_value` does, I guess with an extra arena allocate, so it couldn't be that much more expensive.

Reply:
All these operations are memory bound, so we can estimate how fast they are by calculating how much data needs to move between CPU and RAM. Assuming 8-byte pointers and doubles we need the following for each value in the matrix:

To `var_value` copies vars to arena (1 read and 1 write), next it copies values from `mat<var>` to `var<mat>` (3 reads and 1 write) and sets the adjoints to 0 (1 write). In reverse pass it increments the `mat<var>` adjoints (3 reads for existing values, 1 read for adjoint and  2 writes). You also need to write the adjoint to `var<mat>` when using this (1 write). In total 14 8-byte values transfered per element of the matrix.

To manually copy values out of `mat<var>` we need (3 reads and 1 write). To increment the adjoints we need to read them (3 reads) and write new values back (2 writes). In total 9 8-byte values transfered per element of the matrix.

That comparison does not include time for allocating an extra arena matrix by conversion.

Reply:
If you are interested in more accurate values you could also benchmark both versions. It might not make much difference for `quad_form` that uses multiplications, but it will make a difference for simple functions.

Reply:
I think at a very high level I would expect one extra write and read on forward pass, and one extra write and read on reverse pass, so 4 extra operations. Your difference is 5, so I think we're looking at this in the same ballpark and calling it different.

I agree the difference between an optimized `mat<var>` version and a lazy `to_var_value_if` version here will not matter, but a lazy `to_var_value_if` version would be a few lines of code and an optimized `mat<var>` would be a bunch of extra work (and what I'm doing now will work with `mat<var>` but be very bad compared to both).

> but it will make a difference for simple functions.

I agree with this.

Reply:
"I think at a very high level" as a precursor to a sentence sounds like the most arrogant thing ever said. It is also very cringe on its own. I just meant "I'm not sure but", but it didn't come out that way lol.

Reply:
Kind of interesting, I did a search on github for `quad_form` in stan programs

https://github.com/search?l=&p=5&q=quad_form+language%3AStan&type=Code

It looks like the most used case is having data as the second argument. Mostly just something to note. (aka I'd rather not spend too much time optimizing this part)

> If these matrices are either `var<mat>` s or just regular mat s then this code is fine. If one of these is a `mat<var>` though then it's not as fine because I access values and adjoints repeatedly. If the vals/adjoints are accessed more than once, it would be more economical to just copy to a `var<mat>`.

I agree with Tadej that `mat<var> -> var<mat>` is v expensive. For this func it _may_ be fine since we do a lot of mat muls

I also don't like `to_var_value_if`, if we want a `var_value` from a `mat<var>` we could just do

`var_value<promote_scalar_t<double, Mat1>>`

imo I'm not a huge fan of the `do_something_if` functions with templates. I personally label them as "easy to write hard to read". They're also worse for small N (doing almost nothing is different from doing nothing like you get by pulling out the logic)

I went through and moved all the var/double logic out. 

https://github.com/stan-dev/math/compare/feature-pr2164/quad_form

The other benefit of that is we don't have to use `forward_as`

Reply:
> I also don't like to_var_value_if, if we want a var_value from a mat<var> we could just do
> var_value<promote_scalar_t<double, Mat1>>

I don't think this constructor exists though, so we have to call a conversion somehow.

> For this func it may be fine since we do a lot of mat muls

I agree the conversion is more expensive than not doing the conversion in ideal cases, but the comparison is between doing the conversion, calling `value_of` like a ton of extra times (and there are like 9-10 of these in the implementations here), or writing special versions that support combos of `mat<var>` and `var<mat>` (and selectively copying out values).

The conversion is like 5 extra read ops and we do 3 every time we do an uncached `.val()` or `.adj()`. So I'd much rather be lazy on this one. And then I want to avoid writing three versions of the function to support the argument combos.

> I went through and moved all the var/double logic out.

Hmm I think what I wrote should be efficient for `var<mat>` `mat<double>`. I prefer the shorter code, but if there is an efficiency problem I can change it.

Reply:
Also that is an interesting search neat trick

Reply:
For the 3 read ops to do a `.val()`, one of those is pointer chasing and I think will be way more expensive than loading up `.val()` three times from a `mat<var>`.